pkg/cmd/util/clientcmd/cached_discovery.go

```diff
@@ -0,0 +1,131 @@
+	kclient "k8s.io/kubernetes/pkg/client/unversioned"
+	"k8s.io/kubernetes/pkg/runtime"
+)
+
+// CachedDiscoveryClient implements the functions that dicovery server-supported API groups,
+// versions and resources.
+type CachedDiscoveryClient struct {
+	kclient.DiscoveryInterface
+
+	// cacheDirectory is the directory where discovery docs are held.  It must be unique per host:port combination to work well.
+	cacheDirectory string
+
+	// ttl is how long the cache should be considered valid
+	ttl time.Duration
+}
+
+// ServerResourcesForGroupVersion returns the supported resources for a group and version.
+func (d *CachedDiscoveryClient) ServerResourcesForGroupVersion(groupVersion string) (*unversioned.APIResourceList, error) {
+	filename := filepath.Join(d.cacheDirectory, groupVersion, "serverresources.json")
+	cachedBytes, err := d.getCachedFile(filename)
+	// don't fail on errors, we either don't have a file or won't be able to run the cached check. Either way we can fallback.
+	if err == nil {
+		cachedResources := &unversioned.APIResourceList{}
+		if err := runtime.DecodeInto(kapi.Codecs.UniversalDecoder(), cachedBytes, cachedResources); err == nil {
+			return cachedResources, nil
+		}
+
+	}
+
+	liveResources, err := d.DiscoveryInterface.ServerResourcesForGroupVersion(groupVersion)
+	if err != nil {
+		return liveResources, err
+	}
+
+	d.writeCachedFile(filename, liveResources)
+
+	return liveResources, nil
+}
+
+// ServerResources returns the supported resources for all groups and versions.
+func (d *CachedDiscoveryClient) ServerResources() (map[string]*unversioned.APIResourceList, error) {
+	apiGroups, err := d.ServerGroups()
+	if err != nil {
+		return nil, err
+	}
+	groupVersions := kclient.ExtractGroupVersions(apiGroups)
+	result := map[string]*unversioned.APIResourceList{}
+	for _, groupVersion := range groupVersions {
+		resources, err := d.ServerResourcesForGroupVersion(groupVersion)
+		if err != nil {
+			return nil, err
+		}
+		result[groupVersion] = resources
+	}
+	return result, nil
+}
+
+func (d *CachedDiscoveryClient) ServerGroups() (*unversioned.APIGroupList, error) {
+	filename := filepath.Join(d.cacheDirectory, "servergroups.json")
+	cachedBytes, err := d.getCachedFile(filename)
+	// don't fail on errors, we either don't have a file or won't be able to run the cached check. Either way we can fallback.
+	if err == nil {
+		cachedGroups := &unversioned.APIGroupList{}
+		if err := runtime.DecodeInto(kapi.Codecs.UniversalDecoder(), cachedBytes, cachedGroups); err == nil {
+			return cachedGroups, nil
+		}
+
+	}
+
+	liveGroups, err := d.DiscoveryInterface.ServerGroups()
+	if err != nil {
+		return liveGroups, err
+	}
+
+	d.writeCachedFile(filename, liveGroups)
+
+	return liveGroups, nil
+}
+
+func (d *CachedDiscoveryClient) getCachedFile(filename string) ([]byte, error) {
+	// don't fail on errors, we either don't have a file or won't be able to run the cached check. Either way we can fallback.
```

Comment:
this comment doesn't make sense here. Maybe add it as a godoc for this function if you want it around. You already include it everywhere getCachedFile is called